routing/chainview/queue.go

```diff
@@ -0,0 +1,148 @@
+package chainview
+
+import "sync"
+
+// blockEventType is the possible types of a blockEvent.
+type blockEventType uint8
+
+const (
+	// connected is the type of a blockEvent representing a block
+	// that was connected to our current chain.
+	connected blockEventType = iota
+
+	// disconnected is the type of a blockEvent representing a
+	// block that is stale/disconnected from our current chain.
+	disconnected
+)
+
+// blockEvent represent a block that was either connected
+// or disconnected from the current chain.
+type blockEvent struct {
+	eventType blockEventType
+	block     *FilteredBlock
+}
+
+// blockEventQueue is an ordered queue for block events sent from a
+// FilteredChainView. The two types of possible block events are
+// connected/new blocks, and disconencted/stale blocks. The
+// blockEventQueue keeps the order of these events intact, while
+// still being non-blocking. This is important in order for the
+// chainView's call to onBlockConnected/onBlockDisconnected to not
+// get blocked, and for the consumer of the block events to always
+// get the events in the correct order.
+type blockEventQueue struct {
```

Comment:
Excellent addition! We'll likely re-use portions of this abstraction once we also make the `ChainNotifier` properly re-org aware.